File: backend/llm/openrouter.py
from __future__ import annotations

import requests

from ..config import settings
from .base import LLMResult

_BASE = "https://openrouter.ai/api/v1/chat/completions"


class OpenRouterProvider:
    name = "openrouter"

    def __init__(self, api_key: str, model: str | None = None) -> None:
        self.api_key = api_key
        self.model = model or settings.default_models["openrouter"]

    def _headers(self) -> dict[str, str]:
        return {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
            "HTTP-Referer": "http://localhost:8000",
            "X-Title": "ADAPT",
        }

    def generate(self, *, system: str, user: str, max_tokens: int = 4096) -> LLMResult:
        body = {
            "model": self.model,
            "messages": [
                {"role": "system", "content": system},
                {"role": "user", "content": user},
            ],
            "max_tokens": max_tokens,
            "temperature": 0.4,
        }
        r = requests.post(_BASE, headers=self._headers(), json=body, timeout=120)
        r.raise_for_status()
        data = r.json()
        text = data["choices"][0]["message"]["content"]
        tokens = (data.get("usage") or {}).get("total_tokens")
        return LLMResult(text=text, model=self.model, provider=self.name, token_count=tokens, raw=data)

    def ping(self) -> tuple[bool, str | None]:
        try:
            self.generate(system="Reply with OK only.", user="ping", max_tokens=8)
            return True, None
        except Exception as e:  # noqa: BLE001
            return False, str(e)

File: server/src/services/llm/openrouter.js
const OPENROUTER_BASE = "https://openrouter.ai/api/v1/chat/completions";

const JSON_FENCE = /```(?:json)?\s*(\{.*?\})\s*```/s;

function coerceToPlanJson(text) {
  text = text.trim();
  let candidate = text;
  const m = JSON_FENCE.exec(text);
  if (m) {
    candidate = m[1];
  } else {
    const first = text.indexOf("{");
    const last = text.lastIndexOf("}");
    if (first !== -1 && last !== -1 && last > first) {
      candidate = text.slice(first, last + 1);
    }
  }
  try {
    const data = JSON.parse(candidate);
    data.recommendations = data.recommendations || [];
    data.plan_steps = data.plan_steps || [];
    data.companion_materials = data.companion_materials || [];
    return data;
  } catch {
    return {
      recommendations: [{ title: "Model returned malformed JSON", body: text.slice(0, 600), tag: "other", sources: [] }],
      plan_steps: [],
      companion_materials: [],
    };
  }
}

class OpenRouterProvider {
  constructor(apiKey, model) {
    this.apiKey = apiKey;
    this.model = model;
    this.name = "openrouter";
  }

  async generate({ system, user, maxTokens = 4096 }) {
    const res = await fetch(OPENROUTER_BASE, {
      method: "POST",
      headers: {
        "Authorization": `Bearer ${this.apiKey}`,
        "Content-Type": "application/json",
        "HTTP-Referer": "http://localhost:3000",
        "X-Title": "ADAPT",
      },
      body: JSON.stringify({
        model: this.model,
        messages: [
          { role: "system", content: system },
          { role: "user", content: user },
        ],
        max_tokens: maxTokens,
        temperature: 0.4,
      }),
      signal: AbortSignal.timeout(120000),
    });

    if (!res.ok) {
      const error = await res.text();
      throw new Error(`OpenRouter error ${res.status}: ${error}`);
    }

    const data = await res.json();
    return {
      text: data.choices[0].message.content,
      model: this.model,
      provider: this.name,
      tokenCount: data.usage?.total_tokens,
      raw: data,
    };
  }

  async ping() {
    try {
      await this.generate({ system: "Reply OK", user: "ping", maxTokens: 8 });
      return [true, null];
    } catch (e) {
      return [false, e.message];
    }
  }
}

module.exports = { OpenRouterProvider, coerceToPlanJson };
